### PR TITLE
[#23] Switch to new contract+address linking endpoint

### LINKED ISSUE #23: Switch address+contract linking to new endpoint
Use the new MultiBaas address+contract linking endpoint, with a sane default but configurable number of blocks to sync from.

### PR DESCRIPTION
**Pull request by shoenseiwaso**
Resolves #23 

- Switch to new contract+address linking endpoint in MultiBaas
- Default to 100 blocks back

### CODE REVIEW

File: src/deploy.ts [hunk 1]
```diff
@@ -318,7 +318,8 @@ export class MBDeployer implements MBDeployerI {
    */
   private async linkContractToAddress(
     contract: MultiBaasContract,
-    address: MultiBaasAddress
+    address: MultiBaasAddress,
+    startingBlock: string
```
Comment (daenamkim) [inline src/deploy.ts:322]: I got 500 status when the block number is 2 (or more) blocks ahead the current block.

_MultiBaas (local)_
```sh
INFO[2022-06-20 12:32:03.070648+0900] API call result: /api/v0/chains/ethereum/addresses/sample_nft/contracts  api_key=2 forwarded-for= internal="not found" message="an internal error occurred" method=POST remoteaddr="127.0.0.1:51718" status=500 user=demo@curvegrid.com
```

_Plugin_
```sh
Error: Request failed with status code 500
    at createError (/Users/daenamkim/Developments/workspace/curvegrid/PUBLIC/hardhat-multibaas-plugin/node_modules/axios/lib/core/createError.js:16:15)
    at settle (/Users/daenamkim/Developments/workspace/curvegrid/PUBLIC/hardhat-multibaas-plugin/node_modules/axios/lib/core/settle.js:17:12)
    at IncomingMessage.handleStreamEnd (/Users/daenamkim/Developments/workspace/curvegrid/PUBLIC/hardhat-multibaas-plugin/node_modules/axios/lib/adapters/http.js:322:11)
    at IncomingMessage.emit (node:events:402:35)
    at endReadableNT (node:internal/streams/readable:1343:12)
    at processTicksAndRejections (node:internal/process/task_queues:83:21) {
  config: {
    transitional: {
      silentJSONParsing: true,
...

```

I found this: 

https://github.com/curvegrid/multibaas/blob/b184a5b43d8d39a9b59b433bbb4491c08309346c/server/app/web/api/v0/addresses-api_test.go#L1397-L1403

it seems intended but what about to add `try/catch` to show error message like `Block does not exist` or `Block should not be bigger than the current block (xx)`

Comment (daenamkim) [inline src/deploy.ts:322]: (FYI, I have used [this sample project](https://github.com/curvegrid/multibaas-sample-hardhat-project) for test.)

According to [here](https://github.com/curvegrid/hardhat-multibaas-plugin/pull/25/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R150), an empty string should disable (I am not sure what disabled message in UI) but it looks event monitor running.

empty-string

And I am not sure what the differences between `undefined`, `"0"`, `"-1000"`, `"12"` in the MultiBaas UI, and all same as below. (Lack of context, please teach me :bow:)

_After deploy()_
undefined,-10000, 0, 15

_After deployProxy()_
image

**Review by daenamkim — COMMENTED** [review]
I have tested and left some questions and comments.

File: src/deploy.ts (hunk 1 above)
Comment (dahu33) [reply src/deploy.ts:322]: My bad, I didn't see this during review. We should change to a 400 error. I will file an issue.

Comment (daenamkim) [reply src/deploy.ts:322]: I have added https://github.com/curvegrid/multibaas/issues/3994 on behalf of you.

File: src/type-extensions.ts [hunk 2]
```diff
@@ -75,6 +75,18 @@ export interface DeployOptions {
...
+
+  /**
+   * The block to start syncing the contract from.
+   * 
+   * empty string: disable the MultiBaas Event Monitor
+   *  0: sync from the first block
+   * <0: sync from this number of blocks prior to the current block
+   * >0: sync from a specific block number
+   * 
+   * Defaults to -100, or 100 blocks prior to the current block.
```
[the remark this answers is not in the document]
Comment (shoenseiwaso) [reply src/type-extensions.ts:87]: Correct

File: src/deploy.ts (hunk 1 above)
Comment (shoenseiwaso) [reply src/deploy.ts:322]: Hmmm...good call out. Not sure if this is because the UI has not been updated to work with the new endpoint yet or not.

**Pull request merged**